### ISSUE TITLE
Dialog: focus is on the wrong component

### ISSUE THREAD

**Issue by Sun8642** [primefaces/primefaces#10392]
### Describe the bug

When you open a new dialog, the focus is automatically set on the close button.

### Reproducer

1. Go to https://www.primefaces.org/showcase/ui/df/basic.xhtml
2. Click on the responsive button to open a new dialog with some inputs

### Expected behavior

The first input should be focused (the 'Firstname' input in this case)

### PrimeFaces edition

Community

### PrimeFaces version

13.0.0

### Theme

All

### JSF implementation

Mojarra

### JSF version

2.3

### Java version

17

### Browser(s)

Chrome

**Comment by melloware** [comment]
Here is a MonkeyPatch. Just add this JS code to your project and can you confirm your issue is fixed?

```js
PrimeFaces.focus = function(id, context) {
    var selector = ':not(:submit):not(:button):input:visible:enabled[name], a:first';

    setTimeout(function() {
        var focusFirstElement = function(elements) {
            if (!elements || elements.length === 0) {
                return;
            }

            // first element could be the dialog close button
            var firstElement = elements.eq(0);
            // loop over elements looking for an input
            var inputs = elements.filter(":input");
            if (inputs.length > 0) {
                firstElement = inputs.eq(0);
            }

            PrimeFaces.focusElement(firstElement);
        };

        if (id) {
            var jq = $(PrimeFaces.escapeClientId(id));

            if (jq.is(selector)) {
                jq.trigger('focus');
            } else {
                focusFirstElement(jq.find(selector))
            }
        } else if (context) {
            focusFirstElement($(PrimeFaces.escapeClientId(context)).find(selector))
        } else {
            focusFirstElement($(selector));
        }
    }, 50);

    // remember that a custom focus has been rendered
    // this avoids to retain the last focus after ajax update
    PrimeFaces.customFocus = true;
};
```

**Comment by jepsar** [comment]
@melloware just for my information: why is the milestone of this issue set to 13.0.1 and not also to 14.0.0?

**Comment by melloware** [comment]
You can ONLY set 1 milestone in GitHub.  SO the assumption is that its always fixed in master and then ported to 13.0.1 branch.

**Comment by melloware** [comment]
Here is my reproducer.  Run `mvn clean jetty:run` to test it?
[pf-9941.zip](https://github.com/primefaces/primefaces/files/12172702/pf-9941.zip)

**Comment by melloware** [comment]
Are you using Dialog Framework?

**Comment by jepsar** [comment]
> You can ONLY set 1 milestone in GitHub.

Oh, that sucks.

**Comment by Sun8642** [comment]
I'm using Dialog Framework, I'm opening the dialogs with the "openDynamic" function from a bean, there is some magic happening

**Comment by Sun8642** [comment]
In PF12, we had no focus at all.

**Comment by melloware** [comment]
OK whew.  Dialog Framework has a `focus` property: https://primefaces.github.io/primefaces/13_0_0/#/core/dialogframework?id=dialog-framework

But I am not sure it works properly because of the IFRAME that dialogs are rendered in in DF..

**Comment by Sun8642** [comment]
Anyway, without using the focus property or the focus component in the dialog framework, the focus shouldn't be on the close button, should it?

**Comment by stolp** [comment]
I actually prefer it that way. Dialogs are typically used to guard destructive operations and this prevents the user from accidentely triggering them.

**Comment by melloware** [comment]
Yeah it looks like this issue is specific to the Dialog Framework because it renders inside an IFRAME.

**Comment by Sun8642** [comment]
Should I rename the issue and replacing "Dialog" by "Dialog Framework"?

**Comment by melloware** [comment]
No this issue still existed with Dialog. Can you open a new issue identical but say Dialog Framework?

**Comment by majkers** [comment]
It is not just a matter of dialog framework. I do not know why someone added  `a:first` to selector. It wasn't there in PF 7 version from which I migrated to 12 and it was working. For my application I have a tabView with some links (anchor hrefs) in each tab header. And now (ie 12 version) the focus is on them and not on input. If a documentation says it focuses on input so why selector searches for anchors?

**Comment by majkers** [comment]
It is working on 13.0.2 but I can not migrate to 13 from 12 because it takes months of tests of my large app. I have to do a monkey patch, but still do not understand idea of selecting a:first when in docs it says that focus is for selecting inputs. Just like it was in 7 version of PF. Above all. It is not just dialogs that are hit with this error. I described you my problem without dialogs. Anyway IMHO its algorith is strange but it is working ok in 13. For anyone using 12 version...well monkey patch is the only cure....

**Comment by melloware** [comment]
@majkers here is why: https://github.com/primefaces/primefaces/issues/10115

It's already fixed in 12.0.5 also according to that ticket?

**Comment by majkers** [comment]
Actually I think it is broken from 12.0.5 because it is the version I am using. Here is the reproducer in 12.0.5. You can see that focus is on links in dialog and in page.
[primefaces-test-master.zip](https://github.com/primefaces/primefaces/files/13175739/primefaces-test-master.zip)

**Comment by majkers** [comment]
Yes. I am not sure weather elite users like me can request this. Only pro can I think...

**Comment by majkers** [comment]
I did https://github.com/primefaces/primefaces/issues/10913
